### PR TITLE
fix race condition in save_params

### PR DESCRIPTION
**Pull request by charles-cooper**
Fix https://github.com/s3ql/s3ql/issues/219

### CODE REVIEW

File: src/s3ql/common.py [hunk 1]
```diff
@@ -546,14 +546,29 @@ def load_params(cachepath):
...
         fh.flush()
         os.fsync(fh.fileno())
 
+    # we need to flush the dirents too.
+    # stackoverflow.com/a/41362774
+    # stackoverflow.com/a/5809073
+    os.rename(tmpname, filename)
+    dirfd = None
+    try:
+        dirfd = os.open(os.path.dirname(filename), O_DIRECTORY)
```
Comment (Nikratio) [inline src/s3ql/common.py:565]: I don't think this is an imported symbol.. this will give you an error.

File: src/s3ql/common.py [hunk 2]
```diff
@@ -546,14 +546,29 @@ def load_params(cachepath):
...
+        os.fsync(dirfd)
+    # suppress errors. directory open/fsync does not exist on windows
```
Comment (Nikratio) [inline src/s3ql/common.py:567]: No need to worry about that, S3QL does not work under Windows.

File: src/s3ql/common.py (hunk 1 above)
Comment (charles-cooper) [reply src/s3ql/common.py:565]: it's here https://github.com/s3ql/s3ql/blob/46b269f6efcd35a89c37c9f261e19fd59f38940a/src/s3ql/common.py#L19

File: src/s3ql/common.py (hunk 2 above)
Comment (charles-cooper) [reply src/s3ql/common.py]: resolved in 46b269f

File: src/s3ql/common.py (hunk 1 above)
Comment (Nikratio) [reply src/s3ql/common.py:565]: That doesn't help. Take a look at the CI results if you don't believe me, your code crashes :-).

**Comment by Nikratio** [comment]
Instead of going through the trouble of writing a new file, renaming it, and syncing, wouldn't it be an easier solution to open the existing file without truncating (so that the write then overwrites the existing data), and then truncating to the new length?

I don't think we need to worry that extending the file will ever fail since it's just a few hundred bytes, but even that could be handled by adding a `os.posix_fallocate` after the open.

File: src/s3ql/common.py (hunk 1 above)
Comment (charles-cooper) [reply src/s3ql/common.py:565]: oh, I thought you were talking about `os`. You're right, `O_DIRECTORY` needs to be qualified.

**Comment by charles-cooper** [comment]
> Instead of going through the trouble of writing a new file, renaming it, and syncing, wouldn't it be an easier solution to open the existing file without truncating (so that the write then overwrites the existing data), and then truncating to the new length?
> 
> I don't think we need to worry that extending the file will ever fail since it's just a few hundred bytes, but even that could be handled by adding a `os.posix_fallocate` after the open.

It might be easier, but I don't think it's correct. Overwriting the file in place could leave the file in a mangled state.

File: src/s3ql/common.py [hunk 3]
```diff
@@ -546,14 +546,27 @@ def load_params(cachepath):
...
+    # we need to flush the dirents too.
+    # stackoverflow.com/a/41362774
+    # stackoverflow.com/a/5809073
+    os.rename(tmpname, filename)
+    dirfd = None
+    try:
+        dirfd = os.open(os.path.dirname(filename), os.O_DIRECTORY)
+        os.fsync(dirfd)
+    finally:
+        if dirfd is not None:
```
Comment (Nikratio) [inline src/s3ql/common.py:568]: If you move the dirfd assignment before the try clause, then you need neither this line nor the `dirfd = None` assignment.

Comment (charles-cooper) [reply src/s3ql/common.py:568]: Good point. Thanks

**Pull request merged**